### PR TITLE
MAINT: stats: Some changes for gumbel_r.fit (and therefore gumbel_l.fit).

### PR DESCRIPTION
**Pull request by WarrenWeckesser**
* Reformulate some expressions in gumbel_r.fit to avoid overflow of
  exponential terms.
* Use root-finding for just a scalar equation instead of simultaneously
  solving two equations.
* Add a test for the `fit` method of `gumbel_l` and `gumbel_r` that has
  a precomputed known solution that was computed using mpmath.

### CODE REVIEW

**Comment by WarrenWeckesser** [comment]
@swallan, the PR for the `gumbel_r` and `gumbel_l` `fit` methods looks good.

A while back you asked about the warnings that occur when the scale is small.  I did some experimenting, and I think a reformulation of the expressions that are summed in `func` can avoid some of those warnings.

In the ratio of the sums, a problem arises when a term in `exp(-data/b)` overflows.  This will happen when an element of `data` is negative and `b` is small enough.  We can avoid that issue by using the same technique that is used to avoid overflow in the `logsumexp` function in `scipy.special`.

The expression has the form

    sum(x[i]*exp(c[i]))
    -------------------
      sum(exp(c[i]))

where I'm writing pseudo-ascii-math, and I use `x` instead of `data` and `c` instead of `-data/b`.  That expression can be rewritten as

    exp(M)*sum(x[i]*exp(c[i] - M))    sum(x[i]*exp(c[i] - M))
    ------------------------------ =  -----------------------
      exp(M)*sum(exp(c[i] - M))         sum(exp(c[i] - M))

where `M = max(c)`.  Now all the terms in the exponents are less than or equal to 0, so we don't get overflow.  If we define `p[i] = exp(c[i] - M)`, the expression is

    sum(x[i]*p[i])
    --------------
      sum(p[i])

That expression is the weighted average of `x`, with weights `p`. We can compute that with `numpy.average`.  Here's a function that does that, taking the arguments of the exponentials as a parameter:

    def _average_with_log_weights(x, logweights):
        x = np.asarray(x)
        logweights = np.asarray(logweights)
        maxlogw = logweights.max()
        weights = np.exp(logweights - maxlogw)
        return np.average(x, weights=weights)

Next, note that the equation for the scale does not depend on the location.  So instead of simulaneously solving for (loc, scale), we can use `optimize.root` to solve the (scalar) first equation for the scale, and then plug the scale into the second equation to get the location.

Finally, note that the expression `np.log(sum_exp_x_b / n)` can be implemented in terms of `logsumexp` as `logsumexp(-data/p) - np.log(b)`. So we won't need to explicitly compute `sum_exp_x_b`, and that avoids more potential overflow.

The changes here implement the above reformulations.

With this version, we can do, for example,

    In [6]: z = np.array([3, 3, 3, 3, 3, 3, 3, 3.00000001])

    In [7]: gumbel_r.fit(z)
    Out[7]: (3.00000000016679, 1.249522263435665e-09) 

or

    In [14]: rng = np.random.default_rng(998877665544332211)

    In [15]: z = rng.gumbel(loc=10, scale=1e-9, size=100)

    In [16]: gumbel_r.fit(z)
    Out[16]: (9.999999999901666, 9.83443065337164e-10)

and not get any warnings.

(Ping @mdhaber, @sethtroisi.)

File: scipy/stats/_continuous_distns.py [hunk 1]
```diff
@@ -2927,6 +2927,14 @@ def _entropy(self, c):
 gompertz = gompertz_gen(a=0.0, name='gompertz')
 
 
+def _average_with_log_weights(x, logweights):
```
Comment (sethtroisi) [inline scipy/stats/_continuous_distns.py:2930]: I verified this returns the same answers as the old code but without errors.

logweights - maxlogw is always <= 0 as mentioned so np.exp(...) never overflows :+1:

**Review by sethtroisi — COMMENTED** [review]
looks good at 2am, but I'll do another pass in the light of the day.

**Comment by swallan** [comment]
Thanks for taking the time do to this - I took a look and it seems to work the exact same.

**Pull request merged**